### PR TITLE
[DRAFT] Add Unit Policy Gradient

### PR DESCRIPTION
**Pull request by simoninithomas**
WIP

### CODE REVIEW

File: units/en/unit4/pg-theorem.mdx [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+
+2. The probability of a trajectory (given that action comes from //(/pi_/theta//)):
+
+<img src="https://huggingface.co/datasets/huggingface-deep-rl-course/course-images/resolve/main/en/unit6/probability.png" alt="Probability"/>
+
+
+So we have:
+
+\\(\nabla_\theta J(\theta) =  \nabla_\theta \sum_{\tau}P(\tau;\theta)R(\tau)\\)
```
Comment (osanseviero) [inline units/en/unit4/pg-theorem.mdx:19]: Very cool section! I saw a couple of broken places that require a change

**Review by osanseviero — COMMENTED** [review]
Very cool content! I just missed the hands on, but all the rest looks great!

File: units/en/unit4/advantages-disadvantages.mdx [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+At this point, you might ask, "but Deep Q-Learning is excellent! Why use policy-gradient methods?". To answer this question, let's study the **advantages and disadvantages of policy-gradient methods**.
+
+## Advantages
+
+There are multiple advantages over value-based methods. Let's see some of them:
+
+### The simplicity of integration
+
+Indeed, **we can estimate the policy directly without storing additional data (action values).**
```
[the remark this answers is not in the document]
Comment (simoninithomas) [reply units/en/unit4/advantages-disadvantages.mdx:11]: You don't like indeed :laughing:

File: units/en/unit4/introduction.mdx [hunk 3]
```diff
@@ -0,0 +1,24 @@
+# Introduction [[introduction]]
+
+  <img src="https://huggingface.co/datasets/huggingface-deep-rl-course/course-images/resolve/main/en/unit6/thumbnail.png" alt="thumbnail"/>
+
+In the last unit, we learned about Deep Q-Learning. In this value-based deep reinforcement learning algorithm, we **used a deep neural network to approximate the different Q-values for each possible action at a state.**
+
+Indeed, since the beginning of the course, we only studied value-based methods, **where we estimate a value function as an intermediate step towards finding an optimal policy.**
```
[the remark this answers is not in the document]
Comment (simoninithomas) [reply units/en/unit4/introduction.mdx:7]: Updated :heavy_check_mark:

File: units/en/unit4/introduction.mdx [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+<img src="https://huggingface.co/datasets/huggingface-deep-rl-course/course-images/resolve/main/en/unit3/link-value-policy.jpg" alt="Link value policy" />
+
+Because, in value-based, policy ** \\(π\\) exists only because of the action value estimates, since the policy is just a function** (for instance, greedy-policy) that will select the action with the highest value given a state.
+
+But, with policy-based methods, we want to optimize the policy directly **without having an intermediate step of learning a value function.**
+
+So today, **we'll learn about policy-based methods, and we'll study a subset of these methods called policy gradient**. Then we'll implement our first policy gradient algorithm called Monte Carlo **Reinforce** from scratch using PyTorch.
+Before testing its robustness using CartPole-v1, and PixelCopter environments.
```
[the remark this answers is not in the document]
Comment (simoninithomas) [reply units/en/unit4/introduction.mdx:16]: Updated :heavy_check_mark:

File: units/en/unit4/what-are-policy-based-methods.mdx [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+<img src="https://huggingface.co/datasets/huggingface-deep-rl-course/course-images/resolve/main/en/unit6/soccer.jpg" alt="Soccer" />
+
+## Value-based, Policy-based, and Actor-critic methods
+
+We studied in the first unit, that we had two methods to find (most of the time approximate) this optimal policy \\(\pi*\\).
+
+- In *value-Based methods*, we learn a value function.
+  - The idea then is that an optimal value function leads to an optimal policy \\(\pi^{*}\\).
+  - Our objective is to **minimize the loss between the predicted and target value** to approximate the true action-value function.
+  - We have a policy, but it's implicit since it **was generated directly from the Value function**. For instance, in Q-Learning, we defined an epsilon-greedy policy.
```
[the remark this answers is not in the document]
Comment (simoninithomas) [reply units/en/unit4/what-are-policy-based-methods.mdx:18]: value function. I updated it :heavy_check_mark:

File: units/en/unit4/what-are-policy-based-methods.mdx [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+- On the other hand, in *policy-based methods*, we directly learn to approximate \\(\pi^{*}\\) without having to learn a value function.
+  - The idea then is **to parameterize the policy**. For instance, using a neural network \\(\pi_\theta\\), this policy will output a probability distribution over actions (stochastic policy).
+
+<img src="https://huggingface.co/datasets/huggingface-deep-rl-course/course-images/resolve/main/en/unit6/stochastic_policy.png" alt="stochastic policy" />
+
+  - Our objective then is **to maximize the performance of the parameterized policy using gradient ascent**.
```
[the remark this answers is not in the document]
Comment (simoninithomas) [reply units/en/unit4/what-are-policy-based-methods.mdx:25]: Added :heavy_check_mark:

[the remark this answers is not in the document]
Comment (simoninithomas) [reply units/en/unit4/what-are-policy-based-methods.mdx:25]: Found a solution but not perfect :heavy_check_mark:

File: units/en/unit4/hands-on.mdx [hunk 7]
```diff
@@ -0,0 +1,1012 @@
...
+    api = HfApi()
+
+    # Step 1: Create the repo
+    repo_url = api.create_repo(
+        repo_id=repo_id,
+        exist_ok=True,
+    )
+
+    # Step 2: Download files
+    repo_local_path = Path(snapshot_download(repo_id=repo_id))
```
[the remark this answers is not in the document]
Comment (osanseviero) [reply units/en/unit4/hands-on.mdx:706]: Something like this would work

```python
def push_to_hub(repo_id, 
                model,
                hyperparameters,
                eval_env,
                video_fps=30,
                local_repo_path="hub"
                ):
  """
  Evaluate, Generate a video and Upload a model to Hugging Face Hub.
  This method does the complete pipeline:
  - It evaluates the model
  - It generates the model card
  - It generates a replay video of the agent
  - It pushes everything to the Hub

  :param repo_id: repo_id: id of the model repository from the Hugging Face Hub
  :param model: the pytorch model we want to save
  :param hyperparameters: training hyperparameters
  :param eval_env: evaluation environment
  :param video_fps: how many frame per seconds to record our video replay 
  :param local_repo_path: where the local repository is
  """

  _, repo_name = repo_id.split("/")
  api = HfApi()
  
  # Step 1: Create the repo
  repo_url = api.create_repo(
        repo_id=repo_id,
        exist_ok=True,
  )

  with tempfile.TemporaryDirectory() as tmpdirname:
    local_directory = Path(tmpdirname)
  
    # Step 2: Save the model
    torch.save(model, local_directory / "model.pt")

    # Step 3: Save the hyperparameters to JSON
    with open(local_directory / "hyperparameters.json", "w") as outfile:
      json.dump(hyperparameters, outfile)
    
    # Step 4: Evaluate the model and build JSON
    mean_reward, std_reward = evaluate_agent(eval_env, 
                                            hyperparameters["max_t"],
                                            hyperparameters["n_evaluation_episodes"], 
                                            model)
    # Get datetime
    eval_datetime = datetime.datetime.now()
    eval_form_datetime = eval_datetime.isoformat()

    evaluate_data = {
          "env_id": hyperparameters["env_id"], 
          "mean_reward": mean_reward,
          "n_evaluation_episodes": hyperparameters["n_evaluation_episodes"],
          "eval_datetime": eval_form_datetime,
    }

    # Write a JSON file
    with open(local_directory / "results.json", "w") as outfile:
        json.dump(evaluate_data, outfile)

    # Step 5: Create the model card
    env_name = hyperparameters["env_id"]
    
    metadata = {}
    metadata["tags"] = [
          env_name,
          "reinforce",
          "reinforcement-learning",
          "custom-implementation",
          "deep-rl-class"
      ]

    # Add metrics
    eval = metadata_eval_result(
        model_pretty_name=repo_name,
        task_pretty_name="reinforcement-learning",
        task_id="reinforcement-learning",
        metrics_pretty_name="mean_reward",
        metrics_id="mean_reward",
        metrics_value=f"{mean_reward:.2f} +/- {std_reward:.2f}",
        dataset_pretty_name=env_name,
        dataset_id=env_name,
      )

    # Merges both dictionaries
    metadata = {**metadata, **eval}

    model_card = f"""
    # **Reinforce** Agent playing **{env_id}**
    This is a trained model of a **Reinforce** agent playing **{env_id}** .
    To learn to use this model and train yours check Unit 4 of the Deep Reinforcement Learning Course: https://huggingface.co/deep-rl-course/unit4/introduction
    """

    readme_path = local_directory / "README.md"
    readme = ""
    if readme_path.exists():
        with readme_path.open("r", encoding="utf8") as f:
          readme = f.read()
    else:
      readme = model_card

    with readme_path.open("w", encoding="utf-8") as f:
      f.write(readme)

    # Save our metrics to Readme metadata
    metadata_save(readme_path, metadata)

    # Step 6: Record a video
    video_path =  local_directory / "replay.mp4"
    record_video(env, model, video_path, video_fps)

    # Step 7. Push everything to the Hub
    api.upload_folder(
          repo_id=repo_id,
          folder_path=local_directory,
          path_in_repo=".",
    )

    print(f"Your model is pushed to the Hub. You can view your model here: {repo_url}")
```

File: units/en/unit4/hands-on.mdx [hunk 8]
```diff
@@ -0,0 +1,1012 @@
...
+    api.upload_folder(
+        repo_id=repo_id,
+        folder_path=repo_local_path,
+        path_in_repo=".",
+    )
+
+    print(f"Your model is pushed to the hub. You can view your model here: {repo_url}")
+```
+
+### .
```
Comment (osanseviero) [inline units/en/unit4/hands-on.mdx:790]: To delete?

File: units/en/unit4/hands-on.mdx [hunk 9]
```diff
@@ -0,0 +1,1012 @@
...
+
+```python
+repo_id = ""  # TODO Define your repo id {username/Reinforce-{model-id}}
+push_to_hub(
+    repo_id,
+    cartpole_policy,  # The model we want to save
+    cartpole_hyperparameters,  # Hyperparameters
+    eval_env,  # Evaluation environment
+    video_fps=30,
+    local_repo_path="hub",
```
Comment (osanseviero) [inline units/en/unit4/hands-on.mdx:830]: `local_repo_path` is not used, I think you need to delete this from `push_to_hub` and the other functions

File: units/en/unit4/hands-on.mdx [hunk 10]
```diff
@@ -0,0 +1,1012 @@
...
+        metrics_id="mean_reward",
+        metrics_value=f"{mean_reward:.2f} +/- {std_reward:.2f}",
+        dataset_pretty_name=env_name,
+        dataset_id=env_name,
+    )
+
+    # Merges both dictionaries
+    metadata = {**metadata, **eval}
+
+    model_card = f"""
```
Comment (osanseviero) [inline units/en/unit4/hands-on.mdx:756]: You need to un-indent the following lines, otherwise it will be indented
See https://huggingface.co/osanseviero/reinforce-cartpole-slow  vs https://huggingface.co/osanseviero/reinforce-cartpole-slow2

Comment (osanseviero) [reply units/en/unit4/hands-on.mdx:756]: (so un-indenting 757-759)

File: units/en/unit4/hands-on.mdx [hunk 11]
```diff
@@ -0,0 +1,1012 @@
...
+Before diving into the notebook, you need to:
+
+🔲 📚 [Study Policy Gradients by reading Unit 4](https://huggingface.co/deep-rl-course/unit4/introduction)
+
+# Let's code Reinforce algorithm from scratch 🔥
+
+
+To validate this hands-on for the certification process, you need to push your trained models to the Hub.
+
+- Get a result of >= 350 for `Cartpole-v1`.
```
[the remark this answers is not in the document]
Comment (simoninithomas) [reply units/en/unit4/hands-on.mdx:86]: ✔️

File: units/en/unit4/hands-on.mdx [hunk 12]
```diff
@@ -0,0 +1,1012 @@
...
+import torch.nn.functional as F
+import torch.optim as optim
+from torch.distributions import Categorical
+
+# Gym
+import gym
+import gym_pygame
+
+# Hugging Face Hub
+from huggingface_hub import notebook_login # To log to our Hugging Face account to be able to upload models to the Hub.
```
[the remark this answers is not in the document]
Comment (simoninithomas) [reply units/en/unit4/hands-on.mdx:169]: ✔️

File: units/en/unit4/hands-on.mdx (hunk 7 above)
[the remark this answers is not in the document]
Comment (simoninithomas) [reply units/en/unit4/hands-on.mdx:699]: Modified

File: units/en/unit4/hands-on.mdx (hunk 9 above)
Comment (simoninithomas) [reply units/en/unit4/hands-on.mdx:819]: Done thanks
